generator/internal/api/service_dependencies.go

```diff
@@ -0,0 +1,140 @@
+// a given service.
+//
+// The function traverses `model` starting from the definition of `serviceID`.
+//   - Any message used by a method of the service is included in the results.
+//   - Any message used by LROs of the service is included in the results.
+//   - The results are recursively scanned searching for any fields of the
+//     messages included above.
+//   - If a nested message is included in the results, then the parent message
+//     is also included (recursively) in the results.
+func FindServiceDependencies(model *API, serviceID string) *ServiceDependencies {
+	service, ok := model.State.ServiceByID[serviceID]
+	if !ok {
+		return &ServiceDependencies{}
+	}
+	state := newFindServicesState(serviceID, model)
+	state.seed(service)
+	state.search()
+	return &ServiceDependencies{
+		Messages: slices.Sorted(maps.Keys(state.Messages)),
+		Enums:    slices.Sorted(maps.Keys(state.Enums)),
+	}
+}
+
+type findServiceState struct {
+	// The message IDs that have already been visited
+	Visited map[string]bool
+	// The enums already included in these results
+	Enums map[string]bool
+	// The messages already included in these results
+	Messages   map[string]bool
+	model      *API
+	candidates []*Message
+	serviceID  string
+}
+
+func newFindServicesState(serviceID string, model *API) *findServiceState {
+	return &findServiceState{
+		Visited:   map[string]bool{},
+		Enums:     map[string]bool{},
+		Messages:  map[string]bool{},
+		model:     model,
+		serviceID: serviceID,
+	}
+}
+
+func (state *findServiceState) seed(service *Service) {
+	for _, method := range service.Methods {
+		state.addCandidate(method.InputTypeID)
+		state.addCandidate(method.OutputTypeID)
+		if method.OperationInfo != nil {
+			state.addCandidate(method.OperationInfo.MetadataTypeID)
+			state.addCandidate(method.OperationInfo.ResponseTypeID)
+		}
+	}
+}
+
+func (state *findServiceState) search() {
+	for len(state.candidates) > 0 {
+		candidate := state.candidates[len(state.candidates)-1]
+		state.candidates = state.candidates[0 : len(state.candidates)-1]
+
+		if _, ok := state.Visited[candidate.ID]; ok {
+			continue
+		}
+		state.Messages[candidate.ID] = true
+		state.recurse(candidate)
+	}
+}
+
+func (state *findServiceState) recurse(msg *Message) {
+	state.Visited[msg.ID] = true
+	for _, field := range msg.Fields {
+		switch field.Typez {
+		case ENUM_TYPE:
+			state.addEnum(field.TypezID)
+		case MESSAGE_TYPE:
+			state.addMessage(field.TypezID)
+		default:
+		}
+	}
+}
+
+func (state *findServiceState) addEnum(id string) {
+	if e, ok := state.model.State.EnumByID[id]; ok {
+		if e.Parent != nil {
+			state.addCandidate(e.Parent.ID)
+		}
+		state.Enums[id] = true
+	}
+}
+
+func (state *findServiceState) addMessage(id string) {
+	state.addCandidate(id)
+	if m, ok := state.model.State.MessageByID[id]; ok {
+		if m.Parent != nil {
+			state.addCandidate(m.Parent.ID)
+		}
+		if !m.IsMap {
```

Comment:
why aren't we handling maps? wouldn't we have to recurse over `m.Fields[0]`, `m.Fields[1]`?